### ISSUE TITLE
MirageOS-xen goes PVH (using an updated mini-os)

### ISSUE THREAD

**Issue by hannesm** [mirage/mirage#1007]
this is an alternative to #998 -- considering that mini-os from xen on master (and their latest release) supports PVH -- and mostly from a mail from @yomimono who worked on this. instead of rebasing our xen onto unikraft, update mini-os and use ocaml-freestanding (not done below AFAICT) instead of mirage-platform (so we've to maintain only a single ocaml build).

the pins needed:

```
io-page.2.0.1               git git+file:///home/user/io-page#for-new-minios
io-page-xen.2.0.1           git git+file:///home/user/io-page#for-new-minios
minios-xen.dev              git git+file:///home/user/minios-xen#upstream-with-make-install
mirage.3.0.8                git git+file:///home/user/mirage#new-minios-old-ocaml
mirage-bootvar-xen.0.5.0    git git+file:///home/user/mirage-bootvar-xen#for-new-minios
mirage-xen.3.0.6            git git+file:///home/user/mirage-xen#gnttab-return-types
mirage-xen-minios.0.9.3     git git+file:///home/user/mirage-xen-minios#extra_deps
mirage-xen-ocaml.3.0.5      git git+file:///home/user/mirage-platform#remove-removed-plat
mirage-xen-posix.3.0.4      git git+file:///home/user/mirage-platform#remove-removed-plat
ocaml-src.4.05.0            git git+file:///home/user/ocaml#printf-debug
ocb-stubblr.0.1.1           git git+file:///home/user/ocb-stubblr#new-minios-old-ocaml
```

notes from mindy:
> And I've uploaded those branches to my forks of the relevant repositories.  Looking at the commit log in minios-xen might be instructive; IIRC I found some huge chunk of code that was TODO or uninitialized or something in the minios-xen PVH path that convinced me that I didn't want to tackle this, but my memory is poor and so is/was my understanding.  The problem may not be anything major at all, or might just be some silly mistake of mine.
>
> The ocaml-src pin is just for some printfs in GC initialization; I've pushed it even though you shouldn't need it.

**Comment by hannesm** [comment]
TL;DR: this got real. `opam repo add mirage-dev https://github.com/mirage/mirage-dev.git#xen-pvh` should result in a opam universe where an updated mini-os is used to construct PVH unikernels (see successful CI run https://travis-ci.org/github/mirage/mirage-dev/builds/665538181).

More concretely, mostly thanks to other people's work (@yomimono @talex5 @marmarek), I opened some PRs to get this to work:
- https://github.com/mirage/mini-os/pull/6
- https://github.com/mirage/mirage-platform/pull/212
- https://github.com/mirage/mirage-bootvar-xen/pull/44
- https://github.com/mirage/mirage-xen/pull/22
- https://github.com/mirage/mirage-console/pull/84
- https://github.com/mirage/mirage/pull/1089

What you can do (@xaki23 @reynir @cfcs @marmarek @talex5 whoever else is interested): do the above opam repo add, configure and compile your favourite unikernel and report whether it actually works (compilation + execution) or not. Also, reviewing the patches is very welcome. I tried to cleanup the mini-os patch series from earlier branches (and may have broken something while doing so).

**Comment by hannesm** [comment]
the above branches work for small unikernels (hello, console, etc. from mirage-skeleton), more complex (bigger?) ones (device-usage/prng, device-usage/http-fetch, device-usage/network) fail to start -- there's not a single line of output on the console -- any idea what could be the reason?

**Comment by marmarek** [comment]
It may be about xenstore page.

**Comment by palainp** [comment]
Actually I can start unikernel with size 5858024 bytes but failed to start unikernel with size 5882592 bytes, the pb may be related with the xen page size as @hannesm and @marmarek suggests ?

**Comment by mato** [comment]
Some progress:

Working from the observation that nothing at all shows up on the console, which suggests that the hypercall page, or other Xen-reserved areas are somehow corrupted, I took a look at the resulting ELF files. Apart from "cosmetic" errors in the linker script (`.text.*` input sections not getting folded into a single `.text` output section, easy fix), I noticed that `libminios.a` is linked last, which results in Mini-OS itself getting placed _after_ the OCaml program in the resulting ELF file.

Given that we've seen problems due to Mini-OS presumably never getting tested with "large" applications, I figured that this might be what is causing the corruption, and switched the link order to link `libminios.a` _first_, i.e. before `main.native.o` and _also_ keeping `-lminios` _last_ to resolve dependencies on functions provided by that archive in the other components of the stack. (_mumbles something about layering violations and ..._)

This gets me a bit further with `device-usage/network` (on Xen 4.11.4 as shipped with Debian 10):

```
MirageOS booting...
Demand map pfns at 100000000000-108000000000.
Initialising console ... done.
gnttab_table mapped at 0x100000000000.
gnttab_stubs.c: initialised mini-os gntmap
2020-05-21 09:01:36 -00:00: INF [net-xen frontend] connect 0
Page fault at linear address feffc000, rip 20b162, regs 0x40fc18, sp 40fcc0, our_sp 0x40fbe0, code 2
Thread: *NONE*
RIP: 0008:[<000000000020b162>]
RSP: 0000:000000000040fcc0  EFLAGS: 00010047
RAX: 0000000000000000 RBX: 00000000feffc000 RCX: 0000000000000001
RDX: 0000000000002001 RSI: 0000000000960680 RDI: 0000000000000000
RBP: 00000000004436b0 R08: 0000000000000000 R09: 00000000001fffcc
R10: 0000000000000000 R11: 000000000040fc70 R12: 0000000000960a30
R13: 0000000000000005 R14: 000000000043d0a8 R15: 0000000000960658
base is 0x4436b0 caller is 0
```

`0x20b162` is inside `camlCstruct__set_char_739`. I need to investigate the Mini-OS memory map more to understand whether or not `0xfeffc000` should be a valid address, and what it should be mapped to.

**Comment by palainp** [comment]
I got the same page fault when playing with qubes-mirage-firewall : 
The line `let qubesDB = DB.connect ~domid:0 () in` leads me to exactly the same address fault : `0xfeffc000`. And the line `let qrexec = RExec.connect ~domid:0 () in` gave me a different bad address : `0x100000000044`.
Actually I'm not sure that this is a kernel size issue as I can run hello world in mirage-skeleton even if I add a very long string to growth the hello kernel size above 6Mb.

**Comment by palainp** [comment]
I'm sorry if my error is only Qubes related, you can kick me out in this case. That said, I got a bit further following the page fault's path : starting by adding a single call like `let qubesDB = DB.connect ~domid:0 () in` in an hello_world unikernel, I went down along the system's calls to a `IO.create()` call which fail inside the `make_unsafe` function in `xenstore/client_lwt/xs_client_lwt.ml`. It may be a bad configuration in my system ?

And the other page failure at `@0x100000000044` (when calling `RExec.connect ~domid:0 ()`) is triggered when calling `Export.share_pages_exn` in `vchan-xen/xen/memory_xen.ml`.

**Comment by palainp** [comment]
It still may be a Qubes related problem, but I can reproduce the page fault on address `0xfeffc000` with this small unikernel :
config.ml :
```
open Mirage
let main =
  foreign
    ~packages:[package "duration"; package "vchan-xen";]
    "Unikernel.Hello" (time @-> job)
let () =
  register "hello" [main $ default_time]
```
and unikernel.ml :
```
open Lwt.Infix
let port = match Vchan.Port.of_string "111" with
| Error _ -> failwith "Failed to parse port 111"
| Ok x -> x
module Hello (Time : Mirage_time.S) = struct
  let start _time =
    let client_t = Vchan_xen.client ~domid:0 ~port () in
(*    let server_t = Vchan_xen.server ~domid:0 ~port () in*)
    let rec loop = function
      | 0 -> Lwt.return_unit
      | n ->
        Logs.info (fun f -> f "hello");
        Time.sleep_ns (Duration.of_sec 1) >>= fun () ->
        loop (n-1)
    in
    loop 4
end
```
This leads me to the faulting access with this commands :
`mirage clean && mirage configure -t qubes && make depend && make && ~/qubes-test-mirage/test-mirage hello.xen mirage-test`
```
.[32;1mMirageOS booting....[0m
Demand map pfns at 100000000000-108000000000.
Initialising console ... done.
gnttab_table mapped at 0x100000000000.
gnttab_stubs.c: initialised mini-os gntmap
Page fault at linear address feffc000, rip 1ab052, regs 0x39fce8, sp 39fd90, our_sp 0x39fcb0, code 2
Thread: *NONE*
RIP: 0008:[<00000000001ab052>] 
RSP: 0000:000000000039fd90  EFLAGS: 00010047
RAX: 0000000000000000 RBX: 00000000feffc000 RCX: 0000000000000000
RDX: 0000000000002001 RSI: 0000000000978828 RDI: 0000000000000000
RBP: 0000000000286130 R08: 0000000000000000 R09: 00000000001fffcc
R10: 0000000000000000 R11: 000000000039fd00 R12: 0000000000978bd8
R13: 00000000005f6880 R14: 00000000190100a8 R15: 0000000000978800
base is 0x286130 caller is 0xfffffffffffffff3
base is 0x286148 caller is 0x600000000000064
base is 0x6e756f665f746f4e GPF rip: 1d4f6f, error_code=0
Thread: *NONE*
RIP: 0008:[<00000000001d4f6f>] 
RSP: 0000:000000000039fca8  EFLAGS: 00010016
RAX: 000000000000001b RBX: 6e756f665f746f4e RCX: 000000000000001b
RDX: 000000000000034f RSI: 000000000039fb24 RDI: 0000000000000004
RBP: 000000000039fce8 R08: 0000000000000270 R09: 000000000000001b
R10: 0000000000000020 R11: 0000000000000000 R12: 00000000feffc000
R13: 000ffffffffff000 R14: 00000000190100a8 R15: 0000000000978800
base is 0x39fce8 caller is 0x190100a8
base is 0x978800 caller is 0x978828
base is 0xc00 caller is 0

39fc90: a8 fc 39 00 00 00 00 00 00 00 00 00 00 00 00 00
39fca0: 6f 4f 1d 00 00 00 00 00 d0 83 2e 00 03 00 00 00
39fcb0: 00 c0 ff fe 00 00 00 00 cc 9d 24 00 00 00 00 00
39fcc0: 00 c0 ff fe 00 00 00 00 30 61 28 00 00 00 00 00

39fcd0: d8 8b 97 00 00 00 00 00 80 68 5f 00 00 00 00 00
39fce0: 67 a0 1d 00 00 00 00 00 00 88 97 00 00 00 00 00
39fcf0: a8 00 01 19 00 00 00 00 80 68 5f 00 00 00 00 00
39fd00: d8 8b 97 00 00 00 00 00 30 61 28 00 00 00 00 00

1d4f50: ef e8 ca fb ff ff 48 8b 5d 20 58 5a 0f 1f 40 00
1d4f60: 48 89 de bf 0e 94 25 00 31 c0 e8 f1 dd ff ff 48
1d4f70: 8b 73 08 bf 1c 94 25 00 31 c0 e8 e1 dd ff ff 48
1d4f80: 8b 1b 48 85 db 75 d9 4c 8b a5 98 00 00 00 49 81
```
The commented line (`Vchan_xen.server ~domid:0 ~port () in`) gives me another page fault address (`0x100000000044`) which is inside the area allocated during the boot process (`Demand map pfns at 100000000000-108000000000.`). Vchan_xen is involved in both page fault access, can it be possible that an update of this part must be necessary in order to use pvh domain ?

**Comment by mato** [comment]
@palainp Thanks for the investigations. I'm working on an alternative approach (ditching Mini-OS entirely in favour of low-level Xen PVH boot code written from scratch), so I suggest you hold off on this for now. I'll update with a pointer to my work once I have verified that at least a PoC works as intended.

**Comment by mato** [comment]
We have decided to ditch Mini-OS entirely and replace it with a new Xen platform stack based on a hybrid approach using Solo5. See #1159 for details and progress. With that in mind, I think we can close this?

**Comment by hannesm** [comment]
what mato wrote
